crypto/mimc7/mimc7.go

```diff
@@ -0,0 +1,96 @@
+package mimc7
+
+import (
+	"encoding/hex"
+	"fmt"
+	"math/big"
+
+	"github.com/arnaucube/go-snark/bn128"
+	"github.com/arnaucube/go-snark/fields"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// MaxFieldVal is the maximum value that a element of a field can have, fitting inside an R Finite Field. This value is equal to 2**253 - 1
+const MaxFieldVal = "0x1fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"
+const SEED = "iden3_mimc"
+
+// RElem is a big.Int of maximum 253 bits
+type RElem *big.Int
+
+//BigIntsToRElems converts from array of *big.Int to array of RElem
+func BigIntsToRElems(arr []*big.Int) ([]RElem, error) {
+	var o []RElem
+	maxFieldBytes, err := hex.DecodeString(MaxFieldVal[2:])
+	if err != nil {
+		return o, err
+	}
+	maxVal := new(big.Int).SetBytes(maxFieldBytes)
+
+	for i, a := range arr {
+		// check if fits in a Field R element
+		if a.Cmp(maxVal) != -1 {
+			return o, fmt.Errorf("element in position %v don't fits in Finite Field over R", i)
+		}
+		o = append(o, a)
+	}
+	return o, nil
+}
+
+// RElemsToBigInts converts from array of RElem to array of *big.Int
+func RElemsToBigInts(arr []RElem) []*big.Int {
+	var o []*big.Int
+	for _, a := range arr {
+		o = append(o, a)
+	}
+	return o
+}
+
+func getConstants(fqR fields.Fq, seed string, nRounds int) ([]*big.Int, error) {
+	var cts []*big.Int
+	cts = append(cts, big.NewInt(int64(0)))
+	c := new(big.Int).SetBytes(crypto.Keccak256([]byte(seed)))
+	for i := 1; i < nRounds; i++ {
+		c = new(big.Int).SetBytes(crypto.Keccak256(c.Bytes()))
+
+		n := fqR.Affine(c)
+		cts = append(cts, n)
+	}
+	return cts, nil
+}
+
+func MIMC7Hash(fqR fields.Fq, xIn, k *big.Int, nRounds int) (*big.Int, error) {
+	cts, err := getConstants(fqR, SEED, nRounds)
```

Comment:
`getConstants` is evaluated in every `MIMC7Hash` call, but there are some parts that are constant, like calculating:
`c := new(big.Int).SetBytes(crypto.Keccak256([]byte(seed)))` where `seed` is `SEED`

Reply:
done